File: 0-99/21-merge-two-sorted-lists.py
# coding: utf-8

# Definition for singly-linked list.
# class ListNode:
#	 def __init__(self, x):
#		 self.val = x
#		 self.next = None

from made_list import ListNode

class Solution:
	# @param two ListNodes
	# @return a ListNode
	def mergeTwoLists(self, l1, l2):
		pos1, pos2 = l1, l2

		if l1 == None:
			return l2
		elif l2 == None:
			return l1
		elif l1.val < l2.val:
			head = ListNode(l1.val)
			pos1 = pos1.next
		else:
			head = ListNode(l2.val)
			pos2 = pos2.next

		pre = head
		while pos1!=None and pos2!=None:
			if pos1.val < pos2.val:
				minNum = pos1.val
				pos1 = pos1.next
			else:
				minNum = pos2.val
				pos2 = pos2.next

			cur = ListNode(minNum)
			pre.next = cur
			pre = cur

		if pos1 == None:
			while pos2 != None:
				cur = ListNode(pos2.val)
				pre.next = cur
				pre = cur
				pos2 = pos2.next
		elif pos2 == None:
			while pos1 != None:
				cur = ListNode(pos1.val)
				pre.next = cur
				pre = cur
				pos1 = pos1.next
		return head

a = Solution()
l1 = ListNode('1->3->5->7')
l2 = ListNode('2->4->6->8')
print a.mergeTwoLists(l1, l2)

File: special-1-to-200/21-merge-two-sorted-lists.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	// 归并排序链表版
	var (
		p1, p2  = list1, list2
		head    = &ListNode{}
		mergedP = head
	)
	for p1 != nil || p2 != nil { // 这个地方不要写错，!(p1 == nil && p2 == nil) 可能会更直观
		var minNode = p1
		if p2 != nil {
			if minNode == nil || p2.Val < minNode.Val {
				minNode = p2
			}
		}
		if minNode == p1 {
			p1 = p1.Next
		} else if minNode == p2 {
			p2 = p2.Next
		}
		mergedP.Next = minNode
		mergedP = minNode
	}
	return head.Next
}
